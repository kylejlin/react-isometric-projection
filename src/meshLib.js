const parseColors = (colors, polygons) => {
  return Array.apply(null, {length: polygons}).map((a, i) => (colors[i] || colors[0]) || 'transparent');
};

const cube = (x, y, z, s, ...c) => {
  return rectPrism(x, y, z, s, s, s, ...c);
};

const polygon = (c, ...vertices) => {
  return [[c, ...vertices]];
};

const rectPrism = (x, y, z, dx, dy, dz, ...c) => {
  const c1 = parseColors(c, 6);
  return [
    ...xzRect(x, y, z, dx, dz, c1[0]),
    ...xyRect(x, y, z, dx, dy, c1[1]),
    ...yzRect(x, y, z, dy, dz, c1[2]),
    ...xyRect(x, y, z + dz, dx, dy, c1[3]),
    ...yzRect(x + dx, y, z, dy, dz, c1[4]),
    ...xzRect(x, y + dy, z, dx, dz, c1[5])
  ];
};

const xyRect = (x, y, z, dx, dy, c) => {
  return [
    [c, [x, y, z], [x + dx, y, z], [x + dx, y + dy, z], [x, y + dy, z]]
  ];
};

const xzRect = (x, y, z, dx, dz, c) => {
  return [
    [c, [x, y, z], [x + dx, y, z], [x + dx, y, z + dz], [x, y, z + dz]]
  ];
};

/* TODO
const xzRectPyramid = (x, y, z, dx, dy, dz, ...c) => {
  const c1 = parseColors(c, 5);
  return [
    ...xzRect(x, y, z, dx, dz, c1[0]),
    [c1[1], [x, y, z], []]
  ];
};
*/

const yzRect = (x, y, z, dy, dz, c) => {
  return [
    [c, [x, y, z], [x, y + dy, z], [x, y + dy, z + dz], [x, y, z + dz]]
  ];
};

export default {
  cube,
  polygon,
  rectPrism,
  xyRect,
  xzRect,
  yzRect
};