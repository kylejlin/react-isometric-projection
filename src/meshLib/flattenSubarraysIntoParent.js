export default parentArr => parentArr.reduce((arr, item) => Array.isArray(item) ? arr.concat(item) : arr.concat([item]), [])