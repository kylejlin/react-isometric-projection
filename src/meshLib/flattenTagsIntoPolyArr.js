const flattenTagsIntoPolyArr = (tags, offset = [0, 0, 0], scale = [1, 1, 1]) => {
  const [ox, oy, oz] = offset
  const [sx, sy, sz] = scale

  return tags.reduce((arr, tag) => {
    if (tag.type === 'POLYGON') {
      return arr.concat([{
        color: tag.color,
        points: tag.points.map(([x, y, z]) => [ox + x * sx, oy + y * sy, oz + z * sz])
      }])
    } else if (tag.type === 'GROUP') {
      const [px, py, pz] = tag.position

      const [gdx, gdy, gdz] = tag.dimensions
      const [gvbx, gvby, gvbz] = tag.viewBoxDimensions
      const [gsx, gsy, gsz] = [gdx / gvbx, gdy / gvby, gdz / gvbz]
      
      return arr.concat(flattenTagsIntoPolyArr(tag.children, [ox + px * sx, oy + py * sy, oz + pz * sz], [gsx, gsy, gsz]))
    } else {
      throw new Error('Illegal tag type: ' + tag.type);
    }
  }, [])
}

export default flattenTagsIntoPolyArr
