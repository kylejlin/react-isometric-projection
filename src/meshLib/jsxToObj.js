export default (type, props, ...children) => {
  if ('string' === typeof type) {
    throw new Error('Illegal JSX in mesh definition: ')
  }
  
  const propsWithChildren = Object.assign({}, props, { children })
  
  return type(propsWithChildren)
}