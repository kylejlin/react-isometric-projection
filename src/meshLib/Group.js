import flattenSubarraysIntoParent from './flattenSubarraysIntoParent'
import parseVector from './parseVector'

export default ({ viewBoxDimensions = '1 1 1', dimensions = '1 1 1', position = '0 0 0', children }) => {
  return {
    type: 'GROUP',
    viewBoxDimensions: parseVector(viewBoxDimensions),
    dimensions: parseVector(dimensions),
    position: parseVector(position),
    children: flattenSubarraysIntoParent(children)
  }
}