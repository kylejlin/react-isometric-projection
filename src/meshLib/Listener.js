import flattenSubarraysIntoParent from './flattenSubarraysIntoParent'

export default ({ listeners, children }) => {
  return {
    type: 'LISTENER',
    listeners,
    children: flattenSubarraysIntoParent(children)
  }
}
