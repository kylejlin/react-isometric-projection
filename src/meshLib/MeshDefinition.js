import flattenTagsIntoPolyArr from './flattenTagsIntoPolyArr'

export default ({ viewBoxDimensions, children }) => ({
  type: 'MESH_DEFINITION',
  viewBoxDimensions,
  polygons: flattenTagsIntoPolyArr(children)
})