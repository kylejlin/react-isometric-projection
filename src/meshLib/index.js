import MeshDefinition from './MeshDefinition'

import Group from './Group'
import Polygon from './Polygon'
import Pyramid from './Pyramid'
import Rect from './Rect'

import jsxToObj from './jsxToObj'

export default {
  MeshDefinition,
  
  Group,
  Polygon,
  Pyramid,
  Rect,
  
  jsxToObj
}