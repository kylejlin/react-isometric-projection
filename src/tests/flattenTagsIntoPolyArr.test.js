import flattenTagsIntoPolyArr from '../meshLib/flattenTagsIntoPolyArr'

import meshLib from '../meshLib/index'
import jsxToObj from '../meshLib/jsxToObj'

/** @jsx jsxToObj **/

const { Group, Polygon } = meshLib

test('does not alter polygon array', () => {
  expect(flattenTagsIntoPolyArr([
    <Polygon color="red" points={['0 0 0', '0 1 0', '0 1 1']} />,
    <Polygon color="green" points={['0 0 0', '1 0 0', '1 1 0']} />
  ])).toMatchSnapshot()
})

test('adjusts polgons in groups correctly', () => {
  expect(flattenTagsIntoPolyArr([
    <Polygon color="red" points={['0 0 0', '0 1 0', '0 1 1']} />,
    <Group position="1 0 1" viewBoxDimensions="2 2 2" dimensions="1 1 1">
      <Polygon color="blue" points={['0 0 0', '1 0 2', '2 0 2']} />
    </Group>
  ])).toMatchSnapshot()
})

test('handles nested groups correctly', () => {
  expect(flattenTagsIntoPolyArr([
    <Polygon color="red" points={['0 0 0', '0 1 0', '0 1 1']} />,
    <Group position="1 0 1" viewBoxDimensions="2 2 2" dimensions="1 1 1">
      <Polygon color="blue" points={['0 0 0', '1 0 2', '2 0 2']} />
      <Group position="1 1 1" viewBoxDimensions="3 3 3" dimensions="2 2 2">
        <Polygon color="yellow" points={['0 0 0', '1 0 0', '3 2 1']} />
      </Group>
    </Group>
  ])).toMatchSnapshot()
})