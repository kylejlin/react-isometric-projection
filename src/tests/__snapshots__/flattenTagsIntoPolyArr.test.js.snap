// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adjusts polgons in groups correctly 1`] = `
Array [
  Object {
    "color": "red",
    "points": Array [
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        1,
        0,
      ],
      Array [
        0,
        1,
        1,
      ],
    ],
  },
  Object {
    "color": "blue",
    "points": Array [
      Array [
        1,
        0,
        1,
      ],
      Array [
        1.5,
        0,
        2,
      ],
      Array [
        2,
        0,
        2,
      ],
    ],
  },
]
`;

exports[`does not alter polygon array 1`] = `
Array [
  Object {
    "color": "red",
    "points": Array [
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        1,
        0,
      ],
      Array [
        0,
        1,
        1,
      ],
    ],
  },
  Object {
    "color": "green",
    "points": Array [
      Array [
        0,
        0,
        0,
      ],
      Array [
        1,
        0,
        0,
      ],
      Array [
        1,
        1,
        0,
      ],
    ],
  },
]
`;

exports[`handles nested groups correctly 1`] = `
Array [
  Object {
    "color": "red",
    "points": Array [
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        1,
        0,
      ],
      Array [
        0,
        1,
        1,
      ],
    ],
  },
  Object {
    "color": "blue",
    "points": Array [
      Array [
        1,
        0,
        1,
      ],
      Array [
        1.5,
        0,
        2,
      ],
      Array [
        2,
        0,
        2,
      ],
    ],
  },
  Object {
    "color": "yellow",
    "points": Array [
      Array [
        1.5,
        0.5,
        1.5,
      ],
      Array [
        2.1666666666666665,
        0.5,
        1.5,
      ],
      Array [
        3.5,
        1.8333333333333333,
        2.1666666666666665,
      ],
    ],
  },
]
`;
